t/90h2olog.t

```diff
@@ -0,0 +1,92 @@
+#!perl
+# DTRACE_TEST=1 to skip to check prereqisites except for OS
+# H2OLOG_DEBUG=1 for more runtime logs
+use strict;
+use warnings FATAL => "all";
+use Net::EmptyPort qw(empty_port);
+use Test::More;
+use JSON;
+use t::Util;
+
+plan skip_all => "h2olog is supported only for Linux"
+    if $^O ne 'linux';
+
+my $h2olog_prog = bindir() . "/h2olog";
+my $client_prog = bindir() . "/h2o-httpclient";
+
+unless ($ENV{DTRACE_TEST})  {
+  plan skip_all => "$h2olog_prog not found"
+      unless -e $h2olog_prog;
+
+  plan skip_all => "$client_prog not found"
+      unless -e $client_prog;
+
+  plan skip_all => 'dtrace support is off'
+      unless server_features()->{dtrace};
+}
+
+my $quic_port = empty_port({
+    host  => "127.0.0.1",
+    proto => "udp",
+});
+
+my $server = spawn_h2o({
+    opts => [qw(--mode=worker)],
+    conf => << "EOT",
+listen:
+  type: quic
+  port: $quic_port
+  ssl:
+    key-file: examples/h2o/server.key
+    certificate-file: examples/h2o/server.crt
+hosts:
+  default:
+    paths:
+      /:
+        file.dir: t/assets/doc_root
+EOT
+});
+
+subtest "h2olog", sub {
+  my $tracer = spawn_h2olog({
+    pid => $server->{pid},
+    args => [$ENV{H2OLOG_DEBUG} ? ("-d") : ()],
+  });
+
+  my $trace = $tracer->get_trace(sub {
+    my ($headers, $body) = run_prog("$client_prog -3 https://127.0.0.1:$quic_port/");
+    like $headers, qr{^HTTP/3 200\n}, "req: HTTP/3";
+    is $body, "hello\n", "req: body";
+  });
+
+  if ($ENV{H2OLOG_DEBUG}) {
+    diag "h2olog output:\n", $trace;
+  }
+
+  ok( (map { decode_json($_) } split /\n/, $trace), "h2olog output is valid JSON Lines");
```

Comment:
I appreciate the effort, but I'm not sure if we need to create a "tracer" object to test things? I tend to think that all we need to do is:
1. spawn h2olog process that writes the output to a temporary file.
2. sleep 1 (or do some intelligent polling to determine when h2olog gets attached)
3. run $client_prog
4. sleep 1 (or do something intelligent)
5. kill h2olog
6. read the trace stored in the temporary file

Things to keep in mind are:
* we cannot use a pipe - because it might get blocked and then h2olog would start dropping events
* we cannot tell when h2olog emits all the events, therefore we need to use sleep (in step 4), or inject an event that acts as a sentry

Reply:
Oh, sure! Just because I started from t/90dtrace.t, but your suggestion makes more sense.

done b1a06a2d57b3a62a9fcfa9afe9c70066422b82c4

Reply:
Thank you for the changes. They look better, though I think there still might be room for improvements.

The proposed design uses callbacks. The problem of callbacks is that they are less flexible compared to objects that have methods. `spawn_server`, `spawn_h2o` return objects, therefore allows each test to construct / destroy the processes in any order, without any dependency.

Would it be possible to change the design of the tracer so that it would be defined as an object, that has method to obtain the results?

Reply:
a1d0e9b Done! How about this?

Reply:
Thanks! I think it's much easier to understand now.